lcls-twincat-motion/Library/POUs/Motion/FB_MotionRequest.TcPOU

```diff
@@ -0,0 +1,154 @@
+    EXEC: UINT := 1;
+    START: UINT := 2;
+    MOVING: UINT := 3;
+    DONE: UINT := 4;
+    STOP: UINT := 5;
+    ERROR: UINT := 6;
+END_VAR
+]]></Declaration>
+    <Implementation>
+      <ST><![CDATA[rtExec(CLK:=bExecute);
+ftExec(CLK:=bExecute);
+rtReset(CLK:=bReset);
+
+// Go back to INIT state on reset
+IF rtReset.Q THEN
+    nState := INIT;
+    nQueuedState := INIT;
+END_IF
+IF rtExec.Q OR ftExec.Q THEN
+	bDone := FALSE;
+END_IF
+
+CASE nState OF
+	// Start by setting everything to a known value
+    INIT:
+        nState := EXEC;
+        bError := FALSE;
+        sErrorMessage := '';
+        bBusy := FALSE;
+        bDone := FALSE;
+	// Normal "waiting for move" state
+    EXEC:
+        bMyMove := FALSE;
+		// Looking for a rising edge on bExecute
+        IF rtExec.Q THEN
+			bDone := FALSE;
+            IF stMotionStage.bExecute THEN
+				// Pick interruption behavior via state machine if a move is in progress
+                CASE enumMotionRequest OF
+                    ENUM_MotionRequest.WAIT:
+                        nState := MOVING;
+                        nQueuedState := START;
+                    ENUM_MotionRequest.INTERRUPT:
+                        nState := STOP;
+                        nQueuedState := START;
+                    ENUM_MotionRequest.ABORT:
+                        nState := EXEC;
+                END_CASE
+            ELSE
+                nState := START;
+            END_IF
+        END_IF
+	// Set the correct values on DUT_MotionStage to start a new absolute move
+    START:
+        bMyMove := TRUE;
+        stMotionStage.bExecute := TRUE;
+        stMotionStage.nCommand := ENUM_EpicsMotorCmd.MOVE_ABSOLUTE;
+        stMotionStage.fPosition := fPos;
+        stMotionStage.fVelocity := fVel;
+        stMotionStage.fAcceleration := fAcc;
+        stMotionStage.fDeceleration := fDec;
+        nState := MOVING;
+        nQueuedState := DONE;
+	// Watch an ongoing move, look for the move to end or requests to stop the move from this FB
+    MOVING:
+        bBusy := stMotionStage.bBusy;
+        ftBusy(CLK:=bBusy);
+        IF ftBusy.Q THEN
+			// This will send us to DONE if this was our move and START if we were waiting for someone else's move
+            nState := nQueuedState;
+        END_IF
+		// Implement stop on falling trigger
+        IF ftExec.Q AND bMyMove THEN
+            nState := STOP;
+            nQueuedState := EXEC;
+        END_IF
+	// Pick out the bDone state and return to waiting
+    DONE:
+        bDone := stMotionStage.bDone;
+        nState := EXEC;
+	// Request a stop and wait for it to happen
+    STOP:
+        bMyMove := TRUE;
+        stMotionStage.bExecute := FALSE;
+        IF NOT stMotionStage.bBusy THEN
+			// Sends us to EXEC if we are stopping our own move, or to START if interrupting someone else's
+            nState := nQueuedState;
+        END_IF
+	// Lock us into the error state until the FB is reset
+    ERROR:
+        bMyMove := FALSE;
```

Comment:
One tiny thing we could add here is:

`stMotionStage.fPosition := stMotionStage.stAxisStatus.fActPosition;`

The reason being that you end up in this state whenever you hit a STO and if you don't reset the setpoint to where you currently are, then once you clear the error and reset the axis through the FB you will continue moving to your original position request, which was likely a bad one. I was adding doing this in `FB_PitchControl`: https://github.com/jsheppard95/lcls-twincat-optics/blob/4b8815ec3ee544e4c4884663dbbe6a87c75d68ab/lcls-twincat-optics/lcls_twincat_optics_plc/POUs/FB_PitchControl.TcPOU#L96

but the timing is funny with the motor decelerating and when I actually call the FB, while the behavior is more regular/expected when added here.

Reply:
It sounds like the real bug here is: `once you clear the error and reset the axis through the FB you will continue moving to your original position request`, which is not intended behavior. It sounds like some work needs to be done in `FB_MotionStage` to properly cancel moves and reset `fPosition` after clearing an error state, rather than just doing it in a request FB

Reply:
E.g. there might be a bug now where resetting an error via EPICS resumes the move... I need to take a look and fix it for all pathways

Reply:
@jsheppard95 I can't recreate this. Am I doing something different than you?

1. Set FB_MotionRequest.bExecute to TRUE with position and velocity picked
2. Confirm move started
3. Set bHardwareEnable to FALSE (STO hit)
4. Confirm 0x4260 error
5. Set bHardwareEnable to TRUE (STO released)
6. Set FB_MotionRequest.bReset to TRUE
7. Set DUT_MotionStage.bReset to TRUE (5 or 6 can be done in either order)
8. Move doesn't automatically resume
9. Set FB_MotionRequest.bReset and bExecute to FALSE
10. Can ask for a new move

The only weird thing is that I realize now that FB_MotionRequest's bReset should probably also reset DUT_MotionStage if its move is what caused the error state.

Reply:
I did not reset fPosition after clearing an error state. It has no affect on anything in the code.

Reply:
I should also add that I cannot replicate the original issue I described using `FB_MotionStage` only nor `FB_PitchControl`. I'm having other issues but not related to what I originally described with motion continuing after the STO is released. I must have been doing something weird...

Reply:
Even if the original issue can't be recreated, I'll put a note in the github issues for setting reasonable values to these parameters after error resets and on init. It is confusing to see stuff like `fVelocity` at 0 and if there is a bug where motion gets restarted it's good to have `fPosition` at a safe value.